### PR TITLE
cjxl_ng: replace gflags

### PR DESCRIPTION
**Pull request by mo271**

### CODE REVIEW

**Comment by mo271** [comment]
In the current cjxl, we also do our own flag parsing, but we basically write directly to a `jxl::CompressParams`. In cjxl we want to use `JxlEncoderFrameSettingsSetOption` to set most of the options. However we want to maintain a distingtion between setting a flag to true or fals and not setting it (implying default values, possibly depending on other settings).
Right now we do this by using `jxl::Override` and tracking if some arguments have been set by watching using variables with names `opt_foo_id`.

File: tools/cjxl_ng.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#ifndef TOOLS_CJXL_H_
+#define TOOLS_CJXL_H_
```
Comment (szabadka) [inline tools/cjxl_ng.h:7]: This header does not seem to be used.

File: tools/cjxl_ng_main.cc [hunk 2]
```diff
@@ -19,233 +19,465 @@
...
+        "[modular encoding] do Squeeze transform, 0=false, "
+        "1=true (default: true if lossy, false if lossless)",
+        &responsive, &ParseSigned, 1);
+
+    cmdline->AddOptionFlag('V', "version",
+                           "Print encoder library version number and exit.",
+                           &version, &SetBooleanTrue, 1);
+    cmdline->AddOptionFlag('\0', "quiet", "Be more silent", &quiet,
+                           &SetBooleanTrue, 1);
+    cmdline->AddOptionValue('\0', "print_profile", "0|1",
```
Comment (szabadka) [inline tools/cjxl_ng_main.cc:380]: I think this can be removed if we are aiming at not using internal library.

File: tools/cjxl_ng_main.cc [hunk 3]
```diff
@@ -381,50 +614,43 @@ jxl::Status GetPixeldata(const jxl::PaddedBytes& image_data,
...
+    fprintf(stdout, "Copyright (c) the JPEG XL Project\n");
+    return jpegxl::tools::CjxlRetCode::OK;
+  }
+
+  if (!args.quiet) {
+    fprintf(stderr, "JPEG XL encoder %s\n", version.c_str());
+  }
+
+  const char* filename_in = args.file_in;
+  const char* filename_out = args.file_out;
```
Comment (szabadka) [inline tools/cjxl_ng_main.cc:642]: we could keep using args.file_in, args.file_out (to be consistent with other args)

File: tools/cjxl_ng_main.cc [hunk 4]
```diff
@@ -19,233 +19,465 @@
...
+
+    cmdline->AddOptionValue(
+        '\0', "intensity_target", "N",
+        "Upper bound on the intensity level present in the image in nits. "
+        "Leaving this set to its default of 0 lets libjxl choose a sensible "
+        "default "
+        "value based on the color encoding.",
+        &intensity_target, &ParseIntensityTarget, 1);
+
+    cmdline->AddOptionValue(
```
Comment (szabadka) [inline tools/cjxl_ng_main.cc:285]: add a TODO to wire this up

File: tools/cjxl_ng_main.cc [hunk 5]
```diff
@@ -19,233 +19,465 @@
...
+        'x', "dec-hints", "key=value",
+        "color_space indicates the ColorEncoding, see Description();\n"
+        "icc_pathname refers to a binary file containing an ICC profile.",
+        &color_hints, &ParseAndAppendKeyValue, 1);
+
+    cmdline->AddOptionValue(
```
Comment (szabadka) [inline tools/cjxl_ng_main.cc:291]: TODO to wire this up

File: tools/cjxl_ng_main.cc [hunk 6]
```diff
@@ -827,18 +1062,18 @@ int main(int argc, char** argv) {
                                            &extra_channel_info);
             enc_status = JxlEncoderSetExtraChannelInfo(jxl_encoder, 0,
                                                        &extra_channel_info);
-            if (JXL_ENC_SUCCESS != enc_status) {
+            if (JXL_ENC_SUCCESS != static_cast<int>(enc_status)) {
```
[the remark this answers is not in the document]
Comment (mo271) [reply tools/cjxl_ng_main.cc:1065]: yes, don't know why it was `jxl::Status` before...

File: tools/cjxl_ng_main.cc [hunk 7]
```diff
@@ -755,17 +991,16 @@ int main(int argc, char** argv) {
       {
         JxlBasicInfo basic_info = ppf.info;
         if (basic_info.alpha_bits > 0) num_alpha_channels = 1;
-        basic_info.intensity_target =
-            static_cast<float>(FLAGS_intensity_target);
+        basic_info.intensity_target = static_cast<float>(args.intensity_target);
```
[the remark this answers is not in the document]
Comment (mo271) [reply tools/cjxl_ng_main.cc:994]: done

File: tools/cjxl_ng_main.cc [hunk 8]
```diff
@@ -729,20 +966,19 @@ int main(int argc, char** argv) {
...
+    if (args.lossless_jpeg && IsJPG(image_data)) {
+      if (!cmdline.GetOption(args.opt_lossless_jpeg_id)->matched()) {
         std::cerr << "Note: Implicit-default for JPEG is lossless-transcoding. "
                   << "To silence this message, set --lossless_jpeg=(1|0)."
                   << std::endl;
       }
-      if (FLAGS_jpeg_store_metadata) {
-        if (JXL_ENC_SUCCESS != JxlEncoderStoreJPEGMetadata(jxl_encoder, true)) {
+      if (args.jpeg_store_metadata) {
+        if (JXL_ENC_SUCCESS != JxlEncoderStoreJPEGMetadata(jxl_encoder, 1)) {
```
[the remark this answers is not in the document]
Comment (mo271) [reply tools/cjxl_ng_main.cc:976]: done

File: tools/cjxl_ng_main.cc [hunk 9]
```diff
@@ -684,30 +916,35 @@ int main(int argc, char** argv) {
...
-                   FLAGS_modular_ma_tree_learning_percent,
+                   args.modular_ma_tree_learning_percent,
                    JXL_ENC_FRAME_SETTING_MODULAR_MA_TREE_LEARNING_PERCENT,
                    [](int32_t x) -> std::string {
-                     return (-1 <= x && x <= 100)
-                                ? ""
-                                : "Invalid --modular_ma_tree_learning_percent. "
-                                  "Valid range is {-1, 0, 1, ..., 100}.\n";
+                     return -1 <= x ? ""
+                                    : "Invalid --modular_palette_colors, must "
```
[the remark this answers is not in the document]
Comment (mo271) [reply tools/cjxl_ng_main.cc:923]: done

File: tools/cjxl_ng_main.cc [hunk 10]
```diff
@@ -491,52 +717,50 @@ int main(int argc, char** argv) {
...
-      bool use_container = FLAGS_container;
+      bool use_container = args.container;
       // TODO(tfish): Set use_container according to need of encoded data.
       // This will likely require moving this piece out of flags-processing.
-      if (FLAGS_strip) {
+      if (args.strip) {
         use_container = false;
       }
-      JxlEncoderUseContainer(jxl_encoder, use_container);
+      JxlEncoderUseContainer(jxl_encoder, static_cast<int>(use_container));
```
[the remark this answers is not in the document]
Comment (mo271) [reply tools/cjxl_ng_main.cc:749]: done

File: tools/cjxl_ng.h (hunk 1 above)
Comment (mo271) [reply tools/cjxl_ng.h:7]: removed (was unintended leftover...)

File: tools/cjxl_ng_main.cc (hunk 3 above)
Comment (mo271) [reply tools/cjxl_ng_main.cc:642]: makes sense, done

File: tools/cjxl_ng_main.cc (hunk 5 above)
Comment (mo271) [reply tools/cjxl_ng_main.cc:291]: done

**Pull request merged**